File: 이성태/WEEK8/두 큐 합 같게 만들기.py
from collections import deque
queue1 = [1, 1]
queue2 = [1, 5]
'''
불능 조건
- (두 리스트의 총 합 // 2) 보다 큰 원소가 리스트에 존재하는 경우 불가함
- 두 큐의 초기 길이는 같기 때문에 위의 조건만 고려하면 됨.
- 총 가능한 순회 횟수는 q 길이의 * 3 - 2
  => 모든 리스트를 n만큼 옮기고 다시 2n만큼 분배하는 최대의 경우가 있어서 총 3n, 리스트에 최소 한 개의 원소가 있어야 하므로 2를 빼줌
필수 조건
- deque 라이브러리 사용할 것.
- 최솟값은 전역 변수로 설정하며, 이 전역변수가 넘어가면 바로 return 하여 시간 복잡도를 줄인다.
'''


def solution(queue1, queue2):
    queue1, queue2 = deque(queue1), deque(queue2)
    min_depth = len(queue1) * 3 - 2
    sum1, sum2 = sum(queue1), sum(queue2)
    depth = 0

    while depth < min_depth:
        try:
            if sum1 > sum2:
                element = queue1.popleft()
                queue2.append(element)
                sum1, sum2 = sum1 - element, sum2 + element
            elif sum1 < sum2:
                element = queue2.popleft()
                queue1.append(element)
                sum1, sum2 = sum1 + element, sum2 - element
            else:
                return depth
        except:
            break
        depth += 1
    else:
        return -1


if __name__ == '__main__':
    print(solution(queue1, queue2))

File: 이성태/WEEK8/두 큐 합 같게 만들기.js
const queue1 = [3, 2, 7, 2]
const queue2 = [4, 6, 5, 1]

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
        this._sum = 0;
    }

    push(data) {
        const node = new Node(data);
        if (this.head == null) {
            this.head = node;
            this.head.next = this.tail;
        } else {
            this.tail.next = node;
        }
        
        this.tail = node;
        this.size += 1;
        this._sum += data;
    }

    length() {
        return this.size;
    }

    sum() {
        return this._sum;
    }

    popLeft() {
        const popedData = this.head.data;
        this.head = this.head.next;
        this.size -= 1;
        this.sum -= popedData;
        return popedData;
    }
}

function convertToQueue(arr) {
    const queue = new Queue();
    for (let i = 0; i < arr.length; i++) {
        queue.push(arr[i])
    }
    return queue
}

function solution(queue1, queue2) {
    queue1 = convertToQueue(queue1);
    queue2 = convertToQueue(queue2);
    const min_depth = queue1.length() * 3 - 2;
    let sum1 = queue1._sum;
    let sum2 = queue2._sum;

    for (let i = 0; i < min_depth; i++) {
        try {
            if (sum1 > sum2) {
                const element = queue1.popLeft();
                queue2.push(element);
                sum1 -= element
                sum2 += element
            } else if (sum1 < sum2) {
                const element = queue2.popLeft();
                queue1.push(element);
                sum1 += element
                sum2 -= element
            } else {
                return i
            }
        } catch {
            break
        }
    }
    return -1
}

console.log(solution(queue1, queue2));